pkg/operator/api/configurations.go

```diff
@@ -25,6 +30,69 @@ import (
 	operatorv1pb "github.com/dapr/dapr/pkg/proto/operator/v1"
 )
 
+const defaultConfig = "daprsystem"
+
+var defaultConfigObjKey = client.ObjectKey{Name: defaultConfig, Namespace: os.Getenv("NAMESPACE")}
+
+func (a *apiServer) getDaprSystemConfiguration(ctx context.Context) (*configurationapi.Configuration, error) {
+	cfg := &configurationapi.Configuration{}
+	// as we're using controller-runtime client, we will be using the cache/watch of it, making this call not expensive and up to date
+	err := a.Client.Get(ctx, defaultConfigObjKey, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("trying to get configuration %s: %w", defaultConfigObjKey.String(), err)
+	}
+	return cfg, nil
+}
+
+// sortMetricsSpec Apply .metrics if set. If not, retain .metric. following
+// https://github.com/dapr/dapr/blob/50757fed0b8545f55714885c97a1543607118c4c/pkg/config/configuration.go#L665 but for the k8s type
+func sortMetricsSpec(c *configurationapi.Configuration) {
+	if c.Spec.MetricsSpec == nil {
+		return
+	}
+
+	if c.Spec.MetricsSpec.Enabled != nil {
+		if c.Spec.MetricSpec == nil {
+			c.Spec.MetricSpec = &configurationapi.MetricSpec{}
+		}
+		c.Spec.MetricSpec.Enabled = c.Spec.MetricsSpec.Enabled
+	}
+
+	if len(c.Spec.MetricsSpec.Rules) > 0 {
+		if c.Spec.MetricSpec == nil {
+			c.Spec.MetricSpec = &configurationapi.MetricSpec{}
+		}
+		c.Spec.MetricSpec.Rules = c.Spec.MetricsSpec.Rules
+	}
+
+	if c.Spec.MetricsSpec.HTTP != nil {
+		if c.Spec.MetricSpec == nil {
+			c.Spec.MetricSpec = &configurationapi.MetricSpec{}
+		}
+		c.Spec.MetricSpec.HTTP = c.Spec.MetricsSpec.HTTP
+	}
+}
+
+func (a *apiServer) overrideConfigDefaults(ctx context.Context, cfg *configurationapi.Configuration) error {
```

Comment:
I'm not understanding what this code does. Why do we apply the daprSystem config to sidecars?

Also, what about non-K8s environments?

Reply:
@ItalyPaleAle The code mainly overrides, if not present on the regular config, config attributes (in this case only one but can be expanded in the future) present on the daprsystem config. Mostly useful for cross cutting concerns where you can define a common config to all.  Now saying this, I'm probably missing the use case of an app without a defined config but for which an operator still want to apply common config setting in the cluster different to dapr internal defaults.

is there something similar to daprsystem config in non-k8s environment? I could look to add to that if there is

Reply:
The daprSystem config is meant to configure the control plane, not individual sidecars.

In non-K8s there's no operator so this code path won't be executed. See the fact that it's not meant to configure individual sidecars.

Reply:
`mtls.enabled` is an example of an existing configuration today which configures sidecars.

Reply:
Sort-of, as that is still something that also configures the control plane and the entire cluster

We don't have an existing case of using daprSystem for each individual sidecar, also because it doesn't work in standalone mode.

Reply:
`mTLS` is always enabled for control plane components in Kubernetes, so I would argue this setting is only for `daprd`s.

Reply:
I like the idea to have a central place to change sidecar config without having to manually override each Deployment with annotation.

Now, there is a different approach to this where we could simply allow a Configuration in the "dapr-system" namespace (or whatever the install NS is) to be tagged as the "default" configuration that the sidecar injector will use if Deployment does snot pick one. The pro is that it will be a much simpler change and easy to understand. The disadvantage is that users can only override their app's configuration with an entire new config instead of a merge.

In summary, my question is only if we want to have a "merge" of the default and app's configuration or to have a complete overwrite.

Reply:
This still doesn't solve the problem of Dapr self-hosted. My understanding is that we the project's goal is to maintain feature parity between the two